refactor for event handling with normal network discovery

joshuef opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 04 Dec 23 13:37 UTC
This pull request includes the following changes:

1. The file `data_payments.rs` has modifications to the `Payment` and `PaymentQuote` structs, adding the `#[debug(skip)]` attribute to certain fields.
2. In the file `main.rs`, there are changes on lines 35 and 325, where an import was removed and a variable name was updated.
3. The file `bootstrap.rs` has changes that remove the `stop_bootstrapping` field and related methods from the `ContinuousBootstrap` struct and implementation.
4. In the file `verify_data_location.rs`, the constant `REVERIFICATION_DELAY` was updated to use a value from the `sn_node` module.
5. The file diff summary for the changes in the node module of the sn_node library includes the addition of a constant and a modified public use statement.
6. The file `wallet.rs` has a modification in the `get_faucet` function, where a println statement was changed to use debug format instead of a string.
7. In the file `swarm.rs`, there are several changes, including the removal of an unused variant and its corresponding implementation, removal of code related to logging, and an updated function signature.
8. The file `api.rs` in the sn_client library has an import statement modified and a code block removed.
9. Changes in the `Client` implementation include modifications to match arms related to `NetworkEvent` and the removal of the `stop_bootstrapping` method.
10. The file `transaction.rs` has changes that remove the `Debug` trait implementation from the `Transaction` struct and add a custom implementation, along with an added `hash` field.
11. The file `quote.rs` has changes including an import statement, a modified method signature, and updated usage.
12. In the file `event.rs` in the sn_networking module, there are additions and refactorings of logging statements, as well as error logging.
13. The file `replication.rs` has changes in the `try_interval_replication` function, including updated import statements, function signature, method usage, and added logging and debugging statements.
<!-- reviewpad:summarize:end -->

joshuef commented:
@Thierry61 Thanks for this. main change seems fine.

You'll need to update the following logs to be correct though: https://github.com/maidsafe/safe_network/pull/1044/files#diff-87e0a1c9b86a90561838367d66d6f23f01f6e2faf23488ad4dc270c2e58c8a13L66-L67

That's looking for a node's `name` as I understand it.

Also, we [follow conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for commit messages to auto generate changelogs for release. So if you could reformat your message to follow that it would be great 🙇 (eg `chore: display binary prefix...`)

Yoga07 commented:
Hi @Thierry61 ,

This refactor changes printing from:
```
2022-02-28T10:32:54.755072Z DEBUG sn_launch_tool: Launching node #3...
Starting logging to directory: "C:\\Users\\yogesh\\.safe/node\\local-test-network\\sn-node-3"
Node PID: 23888, prefix: Prefix(), name: 0f4b97(00001111).., age: 96, connection info:
"127.0.0.1:49364"
"000000.." Node resource usage: Process { memory: 73457, virtual_memory: 4762624, cpu_usage: 4.166667, disk_usage: DiskUsage { total_written_bytes: 258504, written_bytes: 19690, total_read_bytes: 852716, read_bytes: 158 } }
```
to this:
```
2022-02-28T10:18:10.663954Z DEBUG sn_launch_tool: Launching node #14...
Starting logging to directory: "C:\\Users\\yogesh\\.safe/node\\local-test-network\\sn-node-14"
"(0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)" Node resource usage: Process { memory: 79785, virtual_memory: 4741402, cpu_usage: 0.35047743, disk_usage: DiskUsage { total_written_bytes: 5550740, written_bytes: 5349776, total_read_bytes: 855853, read_bytes: 2620 } }
Node PID: 36684, prefix: Prefix(), name: 05f2b2(00000101).., age: 64, connection info:
"127.0.0.1:64759"
```
Is that verbose logging intended? Also FYI, printing the type `Prefix` prints the bits of the xorname prefix eg: `Prefix(01011..)` and strips the trailing zeros.

Thierry61 commented:
> if you could reformat your message to follow that it would be great

I modified it.

If it is still not OK, you should be able to do it yourself, as "Allow edits by maintainers" button is checked.

Thierry61 commented:
> printing the type `Prefix` prints the bits of the xorname prefix and strips the trailing zeros

No, printing the type Prefix doesn’t do this. Previously the name of the prefix (`ctx.prefix().await.name()`) was printed which does this, but is not the prefix (`ctx.prefix().await`)

For whatever reason when this function is called in your test the prefix has 256 bits, which is not a normal condition but is the reality. This reality shouldn’t be hidden by printing the prefix name instead of the usual prefix representation.

Or, you can hide the reality in this very specific condition, but this shouldn’t impact usual prefix representation in normal conditions (and I let you code this specific case!).

Thierry61 commented:
> That's looking for a node's `name` as I understand it.

I don’t understand, I just replaced printing the name of the prefix by printing the prefix.

Please note that I only tested the code with an empty prefix, my network was too small to trigger splits.

joshuef commented:
> I don’t understand, I just replaced printing the name of the prefix by printing the prefix.

Ah sorry, I appear to have grabbed the wrong link there. 🤦 

https://github.com/maidsafe/safe_network/pull/1044/files#diff-87e0a1c9b86a90561838367d66d6f23f01f6e2faf23488ad4dc270c2e58c8a13L66-L67

Here the `prefix` var is used expecting it to be `name`. :| Which yes, We need the variable naming to be updated here as that's obviously wrong.

But basically, updating the `prefix` var to be `prefix` and not `name` will mess with the pending logs. So we just need a `name` var created and used in those logs below too.
----

> I modified it.

It's still missing the scope. (https://www.conventionalcommits.org/en/v1.0.0/#summary)

If you can get that in if there are more changes that'd be great. 
----

>This reality shouldn’t be hidden by printing the prefix name instead of the usual prefix representation.

Yeh that's fair enough 👍 

@Thierry61 Just for clarity, what are you looking to achieve here? What is missing from a log such as `prefix: Prefix(101)` eg?

Thierry61 commented:
> Just for clarity, what are you looking to achieve here?

As mentioned in [this topic](https://safenetforum.org/t/update-24-february-2022/36669/14?u=tfa) I didn’t find a proper way to get current node prefix in the logs:

- “Known prefix after update:” is generated in a loop which means several of them can be listed
- “Section updated: prefix:” is reserved for elders and unavailable for adults
- prefix at the end of “Node resource usage:” isn’t in the right format (000000… instead of ())

This PR is about the third point. I am not sure this is the best way to get the prefix for the dashboard, but anyway the prefix should be displayed using its usual binary representation and not its name.

But if you have a better way to get the prefix of current node I am interested.

Thierry61 commented:
> But basically, updating the `prefix` var to be `prefix` and not `name` will mess with the pending logs

I don't use print_resources_usage flag, so what can be done is to move definition of prefix variable and duplicate it in 2 versions:
- original one (`let prefix: &str = &format!("{}", ctx.prefix().await.name());`) in print_resources_usage branch
- new one in (`let prefix: &str = &format!("({:b})", ctx.prefix().await);`) in else branch

This way your logs (or more exactly your stdout) are not messed with.

But you are aware that prefix name doesn't represent much.

Yoga07 commented:
Hey @Thierry61 ,

By type `Prefix` I meant the struct itself and not the variable `prefix`. So for clarification:
```
println!("{:?}", format!("{:?}", ctx.prefix().await));
prints: "Prefix(0)" 

println!("{:?}", format!("{:?}", ctx.prefix().await.name()));
prints: "000000(00000000).."
```

>This PR is about the third point. I am not sure this is the best way to get the prefix for the dashboard, but anyway the prefix should be displayed using its usual binary representation and not its name.
>But if you have a better way to get the prefix of current node I am interested.

Do the above types of formatting not suffice your requirements? It should be easy to get the only prefix out of `Prefix(xxx)` yes? (i.e the trailing zeros are removed already)

Thierry61 commented:
> Do the above types of formatting not suffice your requirements?

First formatting (`Prefix(0)`) would be OK.

But second one (`000000(00000000)..`) obviously not because `00000000` is different from `0` (prefix depth is not right).

Thierry61 commented:
I committed modification suggested by @Yoga07.

Another remark is the unusual trace statement that doesn’t adhere to the principle of least surprise (prefix is the first argument but appears last in the output).

https://github.com/maidsafe/safe_network/blob/507aa3cf0f23ab149c5865e1d9d0c11ddde1c39e/sn/src/node/logging/mod.rs#L71-L75

Why not a standard trace statement with format string as first argument ?

Yoga07 commented:
> Another remark is the unusual trace statement that doesn’t adhere to the principle of least surprise (prefix is the first argument but appears last in the output).
> 
> https://github.com/maidsafe/safe_network/blob/507aa3cf0f23ab149c5865e1d9d0c11ddde1c39e/sn/src/node/logging/mod.rs#L71-L75
> 
> Why not a standard trace statement with format string as first argument ?

Absolutely. I've opened https://github.com/maidsafe/safe_network/pull/1062 including this change just to have the commits 
 a bit neater. 
Thanks, @Thierry61 :)

joshuef committed:
chore(node): separation of cheap NetworkEvent handling

Prevent undue clones

joshuef committed:
chore: tie verification data test to repl time

joshuef committed:
chore(transfers): tidier debug methods for Transactions

joshuef committed:
chore: log on query

joshuef committed:
chore: log swarm.NetworkInfo

joshuef committed:
chore(networking): remove extended spend wait before verification

joshuef committed:
chore(networking): remove triggered bootstrap slowdown

joshuef committed:
test: tie node reward test to number of data.

Upload test records in parallel
Reduce referividation time

joshuef committed:
chore: FORWARDER_CHOOSING_FACTOR docs

joshuef committed:
chore(networking): dont resort closest peers list

joshuef committed:
chore(network): allow replication even below K_VALUE peers

joshuef committed:
fix(client): dont assume verification is always set w/ VerificationConfig

joshuef committed:
chore: improve Replication debug

joshuef committed:
chore(node): refactor NetworkEvent handling

Remove blocking loop on some tasks

joshuef committed:
test(node): update assert over royalty count

If a repayment happens we may top up, so the count could actually be higher
